neurons/miner.py

```diff
@@ -192,452 +206,520 @@ def __init__(self):
             job_type="mining",
         )
 
+        # Add thread cleanup
+        self._cleanup_threads = set()
+
+        # Add timestamp connection cache
+        self._timestamp_connection = None
+
+    def cleanup(self):
+        """Cleanup method to handle thread termination"""
+        for thread in self._cleanup_threads:
+            if thread.is_alive():
+                thread.join(timeout=1.0)
+        self._cleanup_threads.clear()
+        
+        # Close timestamp connection if it exists
+        if self._timestamp_connection:
+            try:
+                self._timestamp_connection.close()
+                self._timestamp_connection = None
+            except Exception as e:
+                tplr.logger.warning(f"Error closing timestamp connection: {e}")
+
+    async def get_block_timestamp(self, block):
+        """Get timestamp for block with connection reuse and proper cleanup"""
+        connection = None
+        try:
+            # Use existing connection if available
+            if self._timestamp_connection:
+                connection = self._timestamp_connection
+            else:
+                connection = bt.subtensor(config=self.config)
+                self._timestamp_connection = connection
+                
+            return connection.query_module("Timestamp", "Now", block=block).value
+        except Exception as e:
+            tplr.logger.warning(f"Error getting timestamp: {e}")
+            # Close and recreate connection on error
+            if connection and connection == self._timestamp_connection:
+                try:
+                    connection.close()
+                except:
+                    pass
+                self._timestamp_connection = None
+                
+            # Create new connection
+            connection = bt.subtensor(config=self.config)
```

Comment:
Can we use this connection for other parts of the code as well? We have the `self.subtensor` defined in the init so it would make sense to me to use that instead of instantiating a new one?